test/websocket.test.js

```diff
@@ -2763,4 +2765,403 @@ describe('WebSocket', () => {
+        },
+        () => {
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
+            perMessageDeflate: { threshold: 0 }
+          });
+
+          ws.on('open', () => {
+            // Add 200ms delay to decompression:
+            makeDeflateSlow(ws, 200);
+
+            // Client closes after 100ms, before decompression finishes:
+            setTimeout(() => ws.close(1000), 100);
+
+            let receivedMessage = false;
+
+            ws.on('message', () => {
+              receivedMessage = true;
+            });
+
+            ws.on('close', () => {
+              if (receivedMessage) {
+                // The client waited for decompression to finish before processing the server's
+                // close frame response which fires the client's close event.
+                wss.close(done);
+              } else {
+                wss.close();
+                done(
+                  new Error(
+                    'Client did not wait for decompression before close event'
+                  )
+                );
+              }
+            });
+          });
+        }
+      );
+
+      wss.on('connection', (ws) => {
+        ws.send('hello', { compress: true }); // This will take the client 200ms to decompress
+      });
+    });
+
+    it('handles socket shutdown while decompressing incoming frames', (done) => {
+      const wss = new WebSocket.Server(
+        {
+          port: 0,
+          perMessageDeflate: { threshold: 0 }
+        },
+        () => {
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
+            perMessageDeflate: { threshold: 0 }
+          });
+
+          ws.on('open', () => {
+            // Add 200ms delay to decompression:
+            makeDeflateSlow(ws, 200);
+
+            // Client closes after 100ms, before decompression finishes:
+            setTimeout(() => {
+              ws.close(1000);
+              ws._socket.end();
+            }, 100);
+
+            let receivedMessage = false;
+
+            ws.on('message', () => {
+              receivedMessage = true;
+            });
+
+            ws.on('close', () => {
+              if (receivedMessage) {
+                // The clien waited for decompression to finish before processing the server's
+                // close frame response which fires the client's close event.
+                wss.close(done);
+              } else {
+                wss.close();
+                done(
+                  new Error(
+                    'Client did not wait for decompression before close event'
+                  )
+                );
+              }
+            });
+          });
+        }
+      );
+
+      wss.on('connection', (ws) => {
+        ws.send('hello', { compress: true }); // This will take the client 200ms to decompress
+      });
+    });
+
+    it('handles connection closures over both half-open-allowed sockets', (done) => {
```

Comment:
I think this and the next tests are not very useful because no data is buffered/compressed/decompressed. They are not testing edge cases.